sdk/core/azure-core/src/io/body_stream.cpp

```diff
@@ -82,54 +86,58 @@ int64_t MemoryBodyStream::OnRead(uint8_t* buffer, int64_t count, Context const&
   return copy_length;
 }
 
-#if defined(AZ_PLATFORM_POSIX)
-int64_t FileBodyStream::OnRead(uint8_t* buffer, int64_t count, Azure::Core::Context const& context)
+#if defined(AZ_PLATFORM_WINDOWS)
+#pragma warning(push)
+// warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead.
+#pragma warning(disable : 4996)
+#endif
+
+FileBodyStream::FileBodyStream(const std::string& filename) : m_offset(0)
 {
-  (void)context;
-  auto result = pread(
-      this->m_fd,
-      buffer,
-      std::min(count, this->m_length - this->m_offset),
-      this->m_baseOffset + this->m_offset);
+  m_fileStreamHolder = fopen(filename.c_str(), "rb");
+  if (m_fileStreamHolder.GetValue() == nullptr)
+  {
+    throw std::runtime_error("Failed to open file for reading.");
+  }
 
-  if (result < 0)
+#if defined(AZ_PLATFORM_WINDOWS)
+  m_filehandle = (HANDLE)_get_osfhandle(_fileno(m_fileStreamHolder.GetValue()));
+
+  LARGE_INTEGER fileSize;
+  if (!GetFileSizeEx(m_filehandle, &fileSize))
   {
-    throw std::runtime_error("Reading error. (Code Number: " + std::to_string(errno) + ")");
+    throw std::runtime_error("Failed to get size of file.");
   }
+  m_length = fileSize.QuadPart;
 
-  this->m_offset += result;
-  return result;
-}
-#elif defined(AZ_PLATFORM_WINDOWS)
-int64_t FileBodyStream::OnRead(uint8_t* buffer, int64_t count, Azure::Core::Context const& context)
-{
-  (void)context;
-  DWORD numberOfBytesRead;
-  auto o = OVERLAPPED();
-  o.Offset = static_cast<DWORD>(this->m_baseOffset + this->m_offset);
-  o.OffsetHigh = static_cast<DWORD>((this->m_baseOffset + this->m_offset) >> 32);
-
-  auto result = ReadFile(
-      this->m_hFile,
-      buffer,
-      // at most 4Gb to be read
-      static_cast<DWORD>(std::min(
-          static_cast<uint64_t>(0xFFFFFFFFUL),
-          static_cast<uint64_t>(std::min(count, (this->m_length - this->m_offset))))),
-      &numberOfBytesRead,
-      &o);
-
-  if (!result)
+  m_parallelBodyStream = new Internal::ParallelFileBodyStream(m_filehandle, 0, m_length);
```

Comment:
Why raw pointer? raw pointer is error-prone. We have to be very careful about copy constructor, move constructor, copy assignment operator, move assignment operator and destructor.

If you want multiple `FileBodyStream` to be able to share the same underlying file handle, use `std::shared_ptr`.
If you don't want to, use `std::unique_ptr`. Struct or class with `unique_ptr` as member is not copiable, so `FileBodyStream` won't be copiable, then we don't need to worry about deallocation.

Reply:
I started with that. The problem with unique_ptr is that it hurts usability of the type a little bit.
It won't allow creation of objects like this:
```C++
auto stream = FileBodyStream("filename");
// Only this will work
FileBodyStream stream("filename");
```

Reply:
> I started with that. The problem with unique_ptr is that it hurts usability of the type a little bit.
> It won't allow creation of objects like this:
> 
> ```c++
> // Not allowed
> auto stream = FileBodyStream("filename");
> // Only this will work
> FileBodyStream stream("filename");
> ```

for unique_ptr you need: 
`auto stream = std::move(FileBodyStream("filename"));`
or
`auto stream = std::make_unique<FileBodyStream>("filename")`

Which is fine because that's expected-moving semantics for a unique_ptr<>

Reply:
yes, remove the `new`  ,  We don't like `new` in this house hahahha

Reply:
> Is there a way to allow regular object construction but still get automatic lifetime management?

Use a `const&` or `&` as the field for the class instead of a pointer.
Then make the main cttor to take a `&&` and also the copy cttor,  Then the object can be constructed from rvalues (temp instances).

But now conceptually you are not owning the stream, you are using some reference to move-construct one for you.

The thing you are trying to save for end-users (the moving semantics) is one of the main features of modern C++.  It exists since C++ 11, so, eleven years would back you up here =D.  Taking unique_ptr as input in cttor, for instance, describes the intention of that class about ownership

Reply:
> for unique_ptr you need:
auto stream = std::move(FileBodyStream("filename"));
or
auto stream = std::make_unique<FileBodyStream>("filename")

> Which is fine because that's expected-moving semantics for a unique_ptr<>


This is wrong. `FileBodyStream("filename")` returns an rvalue. So if `FileBodyStream` supports move construcotr, `auto f = FileBodyStream("filename");` can work.